cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java

```diff
@@ -93,6 +97,35 @@ public ProgramRunner createSparkProgramRunner(Injector injector) {
     }
   }
 
+  /**
+   * Create a new instance of {@link ProgramRunner} for Spark from the given {@link Injector}.
+   *
+   * @param injector The Guice Injector for acquiring CDAP system instances
+   * @param programRunnerClass the class of the ProgramRunner to create
+   * @return a new instance of {@link ProgramRunner}.
+   */
+  private ProgramRunner createSparkProgramRunner(Injector injector, Class<?> programRunnerClass) throws Exception {
+    // We cannot get the SparkProgramRunner instance from the Injector directly as the guice injector will
+    // cache the class, leading to ClassLoader leakage
+    for (Constructor<?> constructor : programRunnerClass.getDeclaredConstructors()) {
+      // Find the @Inject constructor
+      if (!constructor.isAnnotationPresent(Inject.class)) {
+        continue;
+      }
+      constructor.setAccessible(true);
+
+      // Acquire the instances for each parameter for the constructor
+      Type[] paramTypes = constructor.getGenericParameterTypes();
+      Object[] args = new Object[paramTypes.length];
+      int i = 0;
+      for (Type paramType : paramTypes) {
+        args[i++] = injector.getInstance(Key.get(paramType));
```

Comment:
It doesn't because the SparkProgramRunner only takes parameters that come from the CDAP system, hence through the CDAP system classloader. Hum, given that said, probably I should check the classloader of the `paramType` to decide if I can get it from injector directly or have to do instantiation recursively using this method.